rfcs/0136-stabilize-incrementally.md

```diff
@@ -0,0 +1,189 @@
+---
+feature: stabilize-incrementally
+start-date: 2022-09-15
+author: John Ericson
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Stabilize the new command line interface and Flakes in an incremental fashion, leveraging [RFC 134](https://github.com/NixOS/rfcs/pull/134).
+This plan is designed to still efficiently end the current "limbo" era of these unstable yet widely-used features, yet also soothe the longstanding tensions in the Nix community over how we got here.
+
+# Motivation
+[motivation]: #motivation
+
+For the past few years, we've accumulated a massive backlog of unstable features in the form of the new command-line interface and flakes.
+There is now a growing desire to stabilize those features, but as of yet no plan on exactly how to do so.
+The tension between that desire and a lack of a clear plan has loomed over our heads for a while; this RFC aims to dispel it by providing a concrete plan, an plan that hopefully will mitigate the lingering controversies and tensions around Flakes.
+
+## The situation so far
+
+There are a few facts that must be introduced to give context to this situation.
+We can't change these short of a time machine, so we must find a way to live wit them instead.x
+
+### Flakes are very popular
+
+As measured in the community polls, Nix has a lot of new users, dwarfing the number of long-time users.
+Flakes is very popular among these new users.
+Ergo, Flakes is very popular among the Nix community as a whole.
+
+Like it or not, these users have been using Flakes as if it was stable, and we cannot make huge drastic changes that would break their code in hard-to-fix ways.
+
+Many Groups and individuals interested in the continued growth the Nix community see Flakes are popular, an also wish it to be stabilized to attract further still more users, since Flakes are already proven to be popular among users.
+
+### Difficulties in the roll-out
+
+It is undeniable that the roll-out of these new features has been controversial.
+Graham Christensen's blog post [flakes-are-an-obviously-good-thing](https://grahamc.com/blog/flakes-are-an-obviously-good-thing) lays out the procedure issues along the way quite well.
+
+Some people were upset the Flakes RFC was abandoned but the feature merged.
+Other people were fine with experimental features being merged without RFC, but were upset because Flakes never really *felt* experimental.
+Experimental features would be modified a lot subject to community design use, and, most importantly, *discouraged* from being used in production.
+
+### Flakes are criticized for encroaching on other features
```

Comment:
A while back I took the time to read the entire, massive [Flakes RFC discussion](https://github.com/NixOS/rfcs/pull/49).

If you're going to try to summarize that discussion here, you should mention the [problem of the "global registry"](https://github.com/NixOS/rfcs/pull/49#discussion_r304059813), which seemed like a much bigger problem than any "encroachment".  These registries cause power struggles proportional to the success of their community.

Reply:
Yeah maybe. That is however an issue that is sort of orthogonal to the plan here? I hope that successful stabilization of the earlier stuff will generally lower the temperature on debating things with Flakes in come stabilization time *in general*, but I must admit I am not sure how my plan helps with that issue *in particular*.

Maybe though if only a portion of flakes is stabilized that could be drop?

Perhaps I will ask @alyssais what she thinks is best.

Reply:
> That is however an issue that is sort of orthogonal to the plan here?

Presumably not, since the proposal to separate the registry from the flakes RFC was rejected.

> Perhaps I will ask @alyssais what she thinks is best.

She was the most eloquent exponent of that viewpoint but far from the only one.

Reply:
> Presumably not, since the proposal to separate the registry from the flakes RFC was rejected.

I hope it is a reasonable option to decide that Flakes RFC was positioned «everything to stabilise about flakes» and thus separating the registry discussion would be against that approach, but this time we are trying to commit to things in an incremental way with smaller pieces (if only to avoid making GitHub UI utterly unusable) — thus the fundamentals can be stabilised in a few RFCs (like _what exactly_ is pure enough and how to use it reasonably even without flakes), and then even the flakes themselves could be multiple RFCs.

Reply:
I am firm in my view that having an open global flake registry would be a very bad idea, for all the reasons that I've expressed before.  I don't think Flakes can be made non-experimental without first addressing the global registry elephant in the room, because we need to make sure that we avoid (to the extent we still can) the same problem we've encountered here, where people end up relying on behaviour that has never been through RFC, meaning that the result of any future RFC is a foregone conclusion.

Reply:
Could we do a survey about the global registry?
I propose the following question:
What would you like the global registry to be?
- a curated list (status quo)
- nix and nixpkgs only
- empty list

Reply:
Since the RFC (now?) says that stabilising Flakes itself will require further RFCs, the right place to bring up the global registry is on those future RFCs.  So I consider this discussion resolved with respect to this RFC.

Reply:
OK Thanks, will collapse.